compiler/tests-compiler/global_deadcode.ml

```diff
@@ -0,0 +1,75 @@
+open Util
+
+let%expect_test "Eliminates unused functions from functor" =
+  let program =
+    (* Create two instances of Set functor so compiler can't inline one implementation. *)
+    compile_and_parse_whole_program
+      {|
+      module Int_set = Set.Make (Int);;
+      module String_set = Set.Make (String);;
+
+      let int_set = Int_set.singleton 1 in
+      let string_set = String_set.empty in
+      let string_set = String_set.add "hello" string_set in
+      print_endline (string_of_int (Int_set.find 1 int_set) ^ (String_set.find "hello" string_set))
+      |}
+  in
+  (* Expect: compiles functor members used in the program *)
+  print_fun_decl program (Some "find");
+  [%expect
+    {|
+    function find(x, param){
+     var param$0 = param;
+     for(;;){
+      if(! param$0) throw caml_maybe_attach_backtrace(Not_found, 1);
+      var
+       r = param$0[3],
+       v = param$0[2],
+       l = param$0[1],
+       c = caml_call2(Ord[1], x, v);
+      if(0 === c) return v;
+      var r$0 = 0 <= c ? r : l, param$0 = r$0;
+     }
+    }
+    //end |}];
+  print_fun_decl program (Some "add");
+  [%expect {|
+    function add(x, t){
+     if(! t) return [0, 0, x, 0, 1];
+     var r = t[3], v = t[2], l = t[1], c = caml_call2(Ord[1], x, v);
+     if(0 === c) return t;
+     if(0 <= c){var rr = add(x, r); return r === rr ? t : bal(l, v, rr);}
+     var ll = add(x, l);
+     return l === ll ? t : bal(ll, v, r);
+    }
+    //end |}];
+  (* Expect: does not compile unused functor member *)
+  print_fun_decl program (Some "union");
+  [%expect {| not found |}];
+  print_fun_decl program (Some "inter");
+  [%expect {| not found |}]
+
+let%expect_test "Substitutes unused fields with undefined" =
+  let program =
+    compile_and_parse
+      {|
+      let f b x =
+        let t = if b then (1, 2, x) else (3, x, 4) in
+        let (u, _, v) = t in
+        (u, v)
+      in print_int (fst (f true 1) + snd (f false 2))
+      |}
+  in 
+  (* Expect second field in each triple to be replaced by a sentinal variable. *)
+  print_fun_decl program (Some "f");
+  [%expect {|
+    function f(b, x){
+     var t = b ? [0, 1, _c_, x] : [0, 3, _c_, 4], v = t[3], u = t[1];
```

Comment:
in blocks, undefined fields can be omitted 
`var t = b ? [0, 1, , x] : [0, 3, , 4]

Reply:
@micahcantor I would do this in `generate.ml`. You can register in the context which variables are set to `undefined`. I think you can add a case in this match for that:
https://github.com/ocsigen/js_of_ocaml/blob/2b429e532915a82181be28527864971c3d3efc58/compiler/lib/generate.ml#L1323

Reply:
@vouillon  I'm not sure if I understand exactly, since here I think we would need to output some `Javascript.expression` that represents empty strings for the undefined variables, which we can't do. Since there's no "empty" expression, I would think we'd either need to add this to the JS AST or instead do this at the printing/final generation level.

Reply:
In file `javascript.ml`, an `array_literal` is an `element list`, where an `element` can be an `ElementHole` which corresponds to an omitted  undefined field.

Reply:
Done in ca4642592ab4545d8d9fb9384d41f64ec2963f35